묵찌빠게임 [STEP1] Hisop, 쥬봉이

Hi-sop opened the pull request:
@hojun-jo
안녕하세요 moon. 묵찌빠게임 [STEP1] PR 드립니다.
잘 부탁드립니다!

### 고민이 되었던 점
- 과제 하단에 '재귀함수와 순환함수의 장단점 비교' 라는 체크박스가 있어 재귀로 구성해야하나 고민해봤습니다.
- 승패 구분을 위해 `gurad문`을 사용하였습니다.(코드 40번) 이때 최대한 간결하고 가독성이 좋은 방향으로 구현을 하고자 노력한 결과입니다. `조건을 ||연산자로 나열한 부분`에서 가독성이 괜찮은지 그리고 더 간결하거나 성능적으로 더 나은 코드가 있을지 고민이 됩니다.


### 조언을 얻고 싶은 부분
```swift
func startGame(continueRecursive: Bool) -> Bool {
    guard continueRecursive else {
        return false
    }

    print("가위(1), 바위(2), 보(3)! <종료 : 0> : ", terminator: "")

    guard let userChoice = readLine() else {
        print("잘못된 입력입니다. 다시 시도해주세요.")
        return startGame(continueRecursive: true)
    }
    
    switch userChoice {
    case "1", "2", "3":
        let _ = startGame(continueRecursive: makeResult(userChoice: userChoice))
    case "0":
        print("게임 종료")
        break
    default:
        let _ = startGame(continueRecursive: true)
    }
    return false
}

```
- startGame을 재귀로 만들어보았습니다. 이런 느낌으로 쓰는게 맞는지 궁금합니다.

Commit:
feat: 게임 메뉴 출력 및 사용자 입력 받는 함수 생성

jyubong committed:
feat: 승패 판결 함수 추가

Commit:
refactor: 출력문 수정, 컨벤션 수정

jyubong committed:
docs: Update README.md

jyubong committed:
refactor: 결과 비교 함수의 if문 수정

Commit:
refactor: 게임 결과 판별하는 if문을 guard문으로 수정

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,46 @@
 
 import Foundation
```
import Foundation의 필요성에 대해 설명해주실 수 있을까요?

jyubong replied:
`Foundation`은 데이터 저장 및 지속성, 텍스트 처리, 날짜 및 시간 계산, 정렬 및 필터링, 네트워킹 등의 역할을 하는 프레임워크입니다.
현재 코드에서는 swift에서 제공하는 기능만으로도 코드 구현이 가능하여 해당 import는 제거 후 필요시 추가하여 사용하겠습니다.
[swift 공식문서 Foundation](https://developer.apple.com/documentation/foundation)

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,46 @@
+    }
+}
 
+func makeResult(userChoice: String) -> Bool {
```
네이밍은 항상 어려운 부분이죠😅
게임의 결과를 정하는 함수는 맞지만 Bool 타입을 반환하기 때문에 그에 맞는 이름을 고민해 보시는 것이 좋을 것 같습니다!😀

jyubong replied:
makeResult -> compareWithCompter로 변경해보았습니다.

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,46 @@
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+        }
+    }
+}
```
조언을 얻고 싶은 부분에 적어주신 대로 재귀 함수는 그런 식으로 사용하는 것이 맞습니다👍
재귀 함수로도 시도해보시면 좋을 것 같네요!
그리고 재귀 함수와 순환 함수에 어떤 장단점이 있는지 고민하신 부분을 알려주실 수 있을까요?

Hi-sop replied:
재귀함수는 코드가 간결해지고 변수가 적어진다는 장점을 가지고있으며 단점으로는 오버헤드로 인한 비용 발생, 호출횟수가 많아질 경우 스택오버플로우의 가능성이 있다는 단점이 있습니다.
순환함수는 재귀함수에 비해 상대적으로 코드가 복잡하나 속도가 빠릅니다.

---
저희 코드는 재귀함수로 변경하기로 협의했습니다

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,46 @@
+func makeResult(userChoice: String) -> Bool {
+    let computerChoice: String = String(Int.random(in: 1...3))
+    
+    guard computerChoice != userChoice else {
+        print("비겼습니다!")
+        return true
+    }
+    
+    guard (computerChoice == "1" && userChoice == "3") ||
+            (computerChoice == "2" && userChoice == "1") ||
+            (computerChoice == "3" && userChoice == "2") else {
+        print("이겼습니다!")
+        return false
+    }
+    
+    print("졌습니다!")
+    return false
+}
```
고민이 되었던 점에 적어주신 조건을 나열한 부분은 괄호와 개행으로 충분히 좋은 컨벤션인 것같습니다!👍

개인적으로는 특정 조건이 참일 경우 비김/승리/패배 이런 식으로 가는 것이 로직의 흐름을 읽기에는 편하다고 생각합니다.
이 부분은 취향 차이일 수 있으니 두 분께서 상의하신 후 수정 하셔도 괜찮고, 이대로 가셔도 괜찮습니다!

Hi-sop replied:
상의해본 결과 이대로 가기로 협의했습니다!

hojun-jo commented on RockPaperScissors/RockPaperScissors/main.swift:
```diff
@@ -6,5 +6,46 @@
+            return
+        default:
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+        }
```
switch를 잘 사용하셨네요!👍

hojun-jo reviewed (changes requested):
히섭, 쥬봉이 안녕하세요!
스탭1 진행 수고하셨습니다~
이번 스탭은 잘 하고 계신 것같아 고민하신 부분 위주로 간단하게 리뷰 드렸어요😀

---

### 칭찬 드리고 싶은 점
- 간결하고 가독성 좋은 코드를 위해 노력하신 부분이 보이네요👍
- 전체적으로 네이밍과 컨벤션을 잘 하고 계신 것같습니다👍

---

고민 되어었던 점이나 조언을 얻고 싶은 부분에 대해선 코멘트에 남겨두었습니다!

jyubong committed:
refactor: 게임 메뉴 출력 함수 수정, 결과 비교 함수 이름 변경, import 삭제

startGame함수를 재귀함수로 변경
makeResult -> compareWithComputer 이름 변경
import Foundation 삭제

hojun-jo reviewed (approved):
수고하셨습니다! 다음 스탭도 화이팅하세요~
